src/components/color_picker/color_picker.ts

```diff
@@ -246,50 +319,99 @@ export class ColorPicker extends Component<ColorPickerProps, SpreadsheetChildEnv
     this.props.onColorPicked("");
   }
 
-  setCustomColor(ev: Event) {
-    if (!isColorValid(this.state.currentColor)) {
+  AddCustomColor(ev: Event) {
+    if (!isHSLAValid(this.state.currentHslaColor) || !isColorValid(this.state.customHexColor)) {
       ev.stopPropagation();
       this.state.isCurrentColorInvalid = true;
       return;
     }
-    const color = toHex(this.state.currentColor);
+    ev.stopPropagation();
     this.state.isCurrentColorInvalid = false;
-    this.props.onColorPicked(color);
-    this.state.currentColor = color;
+    this.props.onColorPicked(toHex(this.state.customHexColor));
   }
 
   toggleColorPicker() {
     this.state.showGradient = !this.state.showGradient;
   }
 
-  computeCustomColor(ev: MouseEvent) {
-    this.state.isCurrentColorInvalid = false;
-    this.state.currentColor = computeCustomColor(ev);
+  setCustomGradient({ x, y }: PixelPosition) {
+    // TODORAR: round issues
+    // TODORAR make a converter and inverse converter
+    const offsetX = clip(x, 0, GRADIENT_WIDTH);
+    const offsetY = clip(y, 0, GRADIENT_HEIGHT);
+    const deltaX = offsetX / GRADIENT_WIDTH;
+    const deltaY = offsetY / GRADIENT_HEIGHT;
+    const s = 100 * deltaX;
+    const l = 100 * (1 - deltaY) * (1 - 0.5 * deltaX);
+    this.updateColor({ s, l });
   }
 
-  hideMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "none";
+  setCustomHue(x: Pixel) {
+    const h = (360 * x) / GRADIENT_WIDTH;
+    this.updateColor({ h });
   }
 
-  showMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "block";
+  updateColor(newHsl: Partial<Omit<HSLA, "a">>) {
+    this.state.isCurrentColorInvalid = false;
+    this.state.currentHslaColor = { ...this.state.currentHslaColor, ...newHsl };
+    this.state.customHexColor = hslaToHex(this.state.currentHslaColor);
   }
 
-  moveMagnifier(ev: MouseEvent) {
-    this.state.style.background = computeCustomColor(ev);
-    const shiftFromCursor = 10;
-    this.state.style.left = `${ev.offsetX + shiftFromCursor}px`;
-    this.state.style.top = `${ev.offsetY + shiftFromCursor}px`;
+  dragGradientPointer(ev: MouseEvent) {
+    const initialGradientCoordinates = { x: ev.offsetX, y: ev.offsetY };
+    this.setCustomGradient(initialGradientCoordinates);
+
+    const initialMousePosition = { x: ev.clientX, y: ev.clientY };
+
+    // TODORAR: does not work ,we probably should use the magnifier logic instead
+    // while hiding the pointer
+    // copy what goes on in figure drag & drop
+    const onMouseMove = (ev: MouseEvent) => {
+      const currentMousePosition = { x: ev.clientX, y: ev.clientY };
+      const deltaX = currentMousePosition.x - initialMousePosition.x;
+      const deltaY = currentMousePosition.y - initialMousePosition.y;
+
+      // TODORAR too complicated
+      const currentGradientCoordinates = {
```

Comment:
Maybe defining something like 
```
const fullBorderSize = ITEM_BORDER_WIDTH * 2 + 2;
const maximalCoordinates = {
    x: GRADIENT_WIDTH - fullBorderSize,
    y: GRADIENT_HEIGHT - fullBorderSize
};
```
defining outside of the onMouseMove, to be able to do something like 
`x : clip(initialGradientCoordinates.x + deltaX, 0, maximalCoordinates.x)` ?

But once again I'm not sure it will help

Reply:
I changed the variable definitions and it's now a bit better :)